### PR TITLE
mtd: rename mtd_write_page() -> mtd_write_page_raw(), add high-level mtd_write_page()

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The `mtd_write_page()` function will do a 'raw write' to the underlying memory device.
That means for flash based  storage, it will only do 1 -> 0 bit writes while 0 -> 1 writes are ignored.
This is the behavior that e.g. SPIFFS expects (used there for 'blind writes').

For some applications a more high level view is required where any bit write pattern should result in the requested data being written to the storage device.

For flash based storage, this means a read - modify - write cycle is necessary where a whole flash sector is copied to RAM, modified, erased and re-written.

This PR introduces a `mtd_write_page()` function that will do just that (or fall back to `mtd_write_page_raw()` if the storage supports direct writes, e.g. in the case of EEPROM).

The work area is allocated dynamically once on start-up. This is done because the sector size may not be known at compile-time (see #15617).
To not waste memory when this functionality is not needed, the new function is hidden behind the `mtd_hl_write` pseudo-module.

To properly test this, this also hooks up the AT24MAC EEPROM found on the `same54-xpro` as a second MTD device and fixes some bugs with the `tests/mtd_raw` test that this brought to surface.

### Testing procedure

`tests/mtd_raw` has been extended with a `write_page_hl` command. The new command has also been integrated into the automated `test` command:

```
> test 0 # SPI NOR
2021-01-24 01:14:08,361 #  test 0
2021-01-24 01:14:08,362 # [START]
2021-01-24 01:14:09,122 # [SUCCESS]
> test 1 # AT24MAC EEPROM
2021-01-24 01:14:10,881 #  test 1
2021-01-24 01:14:10,882 # [START]
2021-01-24 01:14:10,929 # [SUCCESS]
```

### Issues/PRs references

`mtd_write_page_hl()` is a terrible name.
I'd much rather rename `mtd_write_page()` ->  `mtd_write_page_raw()`.
But this would be an API change.

### CODE REVIEW

**Comment by dylad** [comment]
> mtd_write_page_hl() is a terrible name.
I'd much rather rename mtd_write_page() -> mtd_write_page_raw().
But this would be an API change.

That's true, this is a terrible name. I am not against an API change for this PR, even if it takes a bit longer to be merged.

**Review by dylad — CHANGES_REQUESTED** [review]
There is something wrong for the new `mtd_write` function
On `SAME54-XPRO` with `tests/mtd_raw`, writing some data to page 1 will result to data write to page 0 + 16 bytes as offset (not 256)

See below log:
```
> 2021-01-26 21:17:21,418 #  main(): This is RIOT! (Version: 2021.04-devel-267-gaf606-mtd_write_page_hl)
2021-01-26 21:17:21,420 # Manual MTD test
2021-01-26 21:17:21,438 # init MTD_0… OK (32 MiB)
2021-01-26 21:17:21,455 # init MTD_1… OK (256 byte)
> erase_sector 0 0
2021-01-26 21:17:26,067 #  erase_sector 0 0
read_page 0 0 0 256
2021-01-26 21:17:30,381 #  read_page 0 0 0 256
2021-01-26 21:17:30,388 # 00000000  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,394 # 00000010  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,401 # 00000020  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,407 # 00000030  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,414 # 00000040  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,420 # 00000050  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,426 # 00000060  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,433 # 00000070  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,439 # 00000080  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,446 # 00000090  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,452 # 000000A0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,458 # 000000B0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,465 # 000000C0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,471 # 000000D0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,478 # 000000E0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:30,484 # 000000F0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
write_page 0 0 0 "HelloTest!"
2021-01-26 21:17:36,153 #  write_page 0 0 0 "HelloTest!"
read_page 0 0 0 256
2021-01-26 21:17:38,778 #  read_page 0 0 0 256
2021-01-26 21:17:38,785 # 00000000  48  65  6C  6C  6F  54  65  73  74  21  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,791 # 00000010  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,798 # 00000020  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,804 # 00000030  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,810 # 00000040  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,817 # 00000050  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,823 # 00000060  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,830 # 00000070  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,836 # 00000080  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,842 # 00000090  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,849 # 000000A0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,855 # 000000B0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,861 # 000000C0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,868 # 000000D0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,874 # 000000E0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:38,881 # 000000F0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
read_page 0 1 0 256
2021-01-26 21:17:44,212 #  read_page 0 1 0 256
2021-01-26 21:17:44,219 # 00000100  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,225 # 00000110  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,232 # 00000120  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,238 # 00000130  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,244 # 00000140  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,251 # 00000150  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,257 # 00000160  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,264 # 00000170  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,270 # 00000180  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,276 # 00000190  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,283 # 000001A0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,289 # 000001B0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,296 # 000001C0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,302 # 000001D0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,308 # 000001E0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:44,315 # 000001F0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
write_page 0 1 0 "HelloTest!"
2021-01-26 21:17:49,745 #  write_page 0 1 0 "HelloTest!"
read_page 0 1 0 256!"
2021-01-26 21:17:51,935 #  read_page 0 1 0 256
2021-01-26 21:17:51,943 # 00000100  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,948 # 00000110  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,955 # 00000120  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,961 # 00000130  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,968 # 00000140  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,974 # 00000150  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,980 # 00000160  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,987 # 00000170  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:51,993 # 00000180  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,000 # 00000190  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,006 # 000001A0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,012 # 000001B0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,019 # 000001C0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,025 # 000001D0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,032 # 000001E0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:17:52,038 # 000001F0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
> read_page 0  0 256
2021-01-26 21:18:20,315 #  read_page 0  0 256
2021-01-26 21:18:20,319 # usage: read_page <dev> <page> <offset> <len>
> read_page 0  0 256
2021-01-26 21:18:22,991 #  read_page 0  0 256
2021-01-26 21:18:22,995 # usage: read_page <dev> <page> <offset> <len>
> read_page 0  0 0 256
2021-01-26 21:18:29,348 #  read_page 0  0 0 256
2021-01-26 21:18:29,355 # 00000000  48  65  6C  6C  6F  54  65  73  74  21  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,361 # 00000010  48  65  6C  6C  6F  54  65  73  74  21  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,368 # 00000020  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,374 # 00000030  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,381 # 00000040  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,387 # 00000050  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,393 # 00000060  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,400 # 00000070  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,406 # 00000080  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,413 # 00000090  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,419 # 000000A0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,425 # 000000B0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,432 # 000000C0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,438 # 000000D0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,445 # 000000E0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2021-01-26 21:18:29,451 # 000000F0  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF

```

See below log

**Comment by benpicco** [comment]
oops, mixed up `pages_per_sector` and `page_size`

**Review by dylad — APPROVED** [review]
ACK. Tested work on SAME54-xpro !
Please squash.
Waiting for second ACK now.

File: drivers/mtd/mtd.c [hunk 1]
```diff
@@ -122,11 +134,49 @@ int mtd_write(mtd_dev_t *mtd, const void *src, uint32_t addr, uint32_t count)
     const uint32_t page_shift = bitarithm_msb(mtd->page_size);
     const uint32_t page_mask = mtd->page_size - 1;
 
-    return mtd_write_page(mtd, src, addr >> page_shift, addr & page_mask, count);
+    return mtd_write_page_raw(mtd, src, addr >> page_shift, addr & page_mask, count);
+}
+
+#ifdef MODULE_MTD_HL_WRITE
```
Comment (maribu) [inline drivers/mtd/mtd.c:140]: Maybe rename the module, now that the function has been renamed?

**Review by maribu — APPROVED** [review]
ACK. Code looks good to me and I trust @dylad 's testing results.

If you want to go for a different module name (IMO matching the function is provides would be nice), squash right in.

File: drivers/mtd/mtd.c (hunk 1 above)
Comment (benpicco) [reply drivers/mtd/mtd.c]: Good idea, changed it.

**Comment by dylad** [comment]
Here we go.

**Pull request merged**